gk/fib.c

```diff
@@ -0,0 +1,1640 @@
+ * Initialize a gateway FIB entry.
+ * @gateway the gateway address informaiton.
+ * @ip_prefix the IP prefix,
+ * for which the gateway is responsible.
+ */
+static struct gk_fib *
+init_gateway_fib_locked(const char *ip_prefix, enum gk_fib_action action,
+	const char *gateway, struct gk_config *gk_conf)
+{
+	int ret, gw_prefix_len, ip_prefix_len;
+	struct ipaddr gw_addr, ip_prefix_addr;
+	struct ether_cache *eth_cache;
+	struct gk_fib *gw_fib, *neigh_fib, *ip_prefix_fib;
+	struct gatekeeper_if *iface;
+	struct neighbor_hash_table *neigh_ht;
+
+	/* Initialize the fib entry for the gateway. */
+	ret = convert_str_to_ip(gateway, &gw_addr);
+	if (ret < 0)
+		return NULL;
+
+	ip_prefix_len = parse_ip_prefix(ip_prefix, &ip_prefix_addr);
+	if (ip_prefix_len < 0)
+		return NULL;
+
+	if (gw_addr.proto != ip_prefix_addr.proto) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to initialize a fib entry for gateway, since the gateway and its responsible grantors have different IP versions.\n");
+		return NULL;
+	}
+
+	if (action == GK_FWD_GATEWAY_FRONT_NET)
+		iface = &gk_conf->net->front;
+	else if (likely(action == GK_FWD_GATEWAY_BACK_NET))
+		iface = &gk_conf->net->back;
+	else {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to initialize a fib entry for gateway, since it has invalid action %d.\n",
+			action);
+		return NULL;
+	}
+
+	/* Find the neighbor FIB entry for this gateway. */
+	neigh_fib = find_fib_entry_for_neighbor_locked(
+		&gw_addr, action, iface, gk_conf);
+	if (neigh_fib == NULL)
+		return NULL;
+
+	/* The gateway FIB entry exists. */
+	if (neigh_fib->action == action)
+		return neigh_fib;
+
+	/* Find the Ethernet cached header entry for this gateway. */
+	neigh_ht = &neigh_fib->u.neigh;
+	eth_cache = neigh_get_ether_cache(neigh_ht, &gw_addr.ip);
+	if (eth_cache == NULL)
+		return NULL;
+
+	if (gw_addr.proto == ETHER_TYPE_IPv4)
+		gw_prefix_len = sizeof(struct in_addr) * 8;
+	else 
+		gw_prefix_len = sizeof(struct in6_addr) * 8;
+
+	/* Add a FIB entry for the gateway. */
+	gw_fib = create_fib_entry_for_gateway_locked(&gw_addr, gw_prefix_len,
+		action, &gw_addr, eth_cache, neigh_ht, iface, gk_conf);
+	if (gw_fib == NULL)
+		return NULL;
+
+	/*
+	 * Add a FIB entry for the IP prefix,
+	 * for which the gateway is responsible.
+	 */
+	ip_prefix_fib = create_fib_entry_for_gateway_locked(
+		&ip_prefix_addr, ip_prefix_len, action, &gw_addr,
+		eth_cache, neigh_ht, iface, gk_conf);
+	if (ip_prefix_fib == NULL)
+		goto del_gw_fib;
+
+	return gw_fib;
+
+del_gw_fib:
+	eth_cache->ref_cnt--;
+	if (eth_cache->ref_cnt == 0)
+		neigh_del_ether_cache(neigh_ht, &gw_addr.ip);
+	initialize_fib_entry(gw_fib);
+
+	return NULL;
+}
+
+static struct gk_fib *
+find_fib_entry_for_grantor_locked(struct ipaddr *gt_addr,
+	int gt_prefix_len, int *gt_fib_present, struct gk_config *gk_conf)
+{
+	int ret;
+	struct gk_fib *gt_prefix_fib;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+
+	if (gt_addr->proto == ETHER_TYPE_IPv4 &&
+			ipv4_if_configured(&gk_conf->net->back)) {
```

Comment:
Maybe it's simpler to leave as it is now, since we need to find the corresponding interface to do the test, which should be the functionality inside the ``add_fib_entry_locked()``.